editor/sidebar/post-taxonomies/tags-selector.js

```diff
@@ -1,36 +1,134 @@
+import { getEditedPostAttribute } from '../../selectors';
+import { editPost } from '../../actions';
+
+const DEFAULT_TAGS_QUERY = {
+	number: 1000,
+	order_by: 'count',
+	order: 'DESC',
+};
+const MAX_TERMS_SUGGESTIONS = 20;
 
 class TagsSelector extends Component {
 	constructor() {
 		super( ...arguments );
-		this.onTokensChange = this.onTokensChange.bind( this );
+		this.onTagsChange = this.onTagsChange.bind( this );
 		this.state = {
-			tokens: [ 'React', 'Vue' ],
+			loading: true,
+			availableTags: [],
+			selectedTags: [],
 		};
 	}
 
-	onTokensChange( value ) {
-		this.setState( { tokens: value } );
+	componentDidMount() {
+		this.fetchTagsRequest = new wp.api.collections.Tags().fetch( DEFAULT_TAGS_QUERY )
+			.done( ( tags ) => {
+				this.setState( {
+					loading: false,
+					availableTags: tags,
+				} );
+				this.updateSelectedTags( this.props.tags );
+			} )
+			.fail( ( xhr ) => {
+				if ( xhr.statusText === 'abort' ) {
+					return;
+				}
+				this.setState( {
+					loading: false,
+				} );
+			} );
+	}
+
+	componentWillUnmount() {
+		if ( this.fetchTagsRequest ) {
+			this.fetchTagsRequest.abort();
+		}
+	}
+
+	componentWillReceiveProps( newProps ) {
+		if ( newProps.tags !== this.props.tags ) {
+			this.updateSelectedTags( newProps.tags );
+		}
+	}
+
+	updateSelectedTags( tags = [] ) {
+		const selectedTags = tags.map( ( tagId ) => {
+			const tagObject = find( this.state.availableTags, ( tag ) => tag.id === tagId );
+			return tagObject ? tagObject.name : '';
+		} );
+		this.setState( {
+			selectedTags,
+		} );
+	}
+
+	onTagsChange( tagNames ) {
+		this.setState( { selectedTags: tagNames } );
+		const newTagNames = tagNames.filter( ( tagName ) =>
+			! find( this.state.availableTags, ( tag ) => tag.name === tagName )
+		);
+		const tagNamesToIds = ( names, availableTags ) => {
+			return names
+				.map( ( tagName ) =>
+					find( availableTags, ( tag ) => tag.name === tagName ).id
+				);
+		};
+
+		if ( newTagNames.length === 0 ) {
+			return this.props.onUpdateTags( tagNamesToIds( tagNames, this.state.availableTags ) );
+		}
+		const createTag = ( tagName ) => new wp.api.models.Tag( { name: tagName } ).save();
+		Promise
+			.all( newTagNames.map( createTag ) )
+			.then( ( newTags ) => {
+				const newAvailableTags = this.state.availableTags.concat( newTags );
+				this.setState( { availableTags: newAvailableTags } );
+				return this.props.onUpdateTags( tagNamesToIds( tagNames, newAvailableTags ) );
+			} );
 	}
 
 	render() {
-		const suggestions = [ 'React', 'Vue', 'Angular', 'Cycle', 'PReact', 'Inferno' ];
+		const { loading, availableTags, selectedTags } = this.state;
+		const tagNames = availableTags.map( ( tag ) => tag.name );
 
 		return (
 			<div className="editor-post-taxonomies__tags-selector">
-				<FormTokenField
-					value={ this.state.tokens }
-					suggestions={ suggestions }
-					onChange={ this.onTokensChange }
-				/>
+				{ loading && <Spinner /> }
```

Comment:
How do you feel about a `disabled` token field instead of a spinner? Is that any better or worse for perceived reactivity?